azure-iot-device/azure/iot/device/provisioning/internal/polling_machine.py

```diff
@@ -0,0 +1,325 @@
+            rid = registration_result.request_id
+            if rid in self._operations:
+                publish_action = self._operations[rid]
+                # If this request was generated by us for a query operation,
+                # then we are getting a response for a query operation and
+                # and we must clear the polling timer before beginning anything
+                if "operationId" in publish_action.publish_topic:
+                    self._polling_timer.cancel()
+                    self._polling_timer = None
+            self._trig_wait(registration_result)
+        elif registration_result.status == "assigned":
+            self._trig_complete(registration_result)
+        else:
+            self._trig_error(registration_result)
+
+    def clear_and_disconnect(self, event_data):
+        self._query_timer = None
+        self._polling_timer = None
+        unsubscribe_action = UnsubscribeAction(
+            constant.SUBSCRIBE_TOPIC_PROVISIONING, callback=self.on_unsubscribe_completed
+        )
+        self._request_response_provider.disconnect(
+            callback=self.on_disconnect_completed, pre_disconnect_actions=[unsubscribe_action]
+        )
+        # TODO : Do I need to wait for disconnect and then call the cancel callback?
+        callback = self._cancel_callback
+
+        if callback:
+            self._cancel_callback = None
+            callback()
+
+    def _call_error(self, event_data):
+        logger.info("Failed register from polling machine")
+        error = event_data.args[0]
+        callback = self._register_callback
+
+        if callback:
+            self._register_callback = None
+            callback(error)
+
+    def _call_complete(self, event_data):
+        logger.info("Complete register from polling machine")
+        registration_result = event_data.args[0]
+        callback = self._register_callback
+
+        if callback:
+            self._register_callback = None
+            callback(registration_result)
+        # TODO : Should disconnect be by default called here ?
+
+    def _set_query_timer(self):
+        def time_up_query():
+            logger.error("Time is up for query timer")
+            # TODO TimeoutError causing problems
+            self._trig_error(ValueError("Time is up for query timer"))
+
+        self._query_timer = Timer(constant.DEFAULT_TIMEOUT_INTERVAL / 1000, time_up_query)
+        self._query_timer.start()
+
+    def _wait_for_interval(self, event_data):
+        def time_up_polling():
+            logger.info(
+                "Done waiting for polling interval of %s ms", str(constant.DEFAULT_POLLING_INTERVAL)
+            )
+
+        self._polling_timer = Timer(constant.DEFAULT_POLLING_INTERVAL / 1000, time_up_polling)
+        logger.info("Waiting for " + str(constant.DEFAULT_POLLING_INTERVAL) + " ms")
+        self._polling_timer.run()  # This is waiting for that polling interval
+        self._trig_poll(event_data)
```

Comment:
>self._trig_poll(event_data) [](start = 8, length = 27)

This is blocking the calling thread while it's waiting.  I don't think you want to do that.  Why not call _trig_poll inside the time_up_polling function and use start() instead of run()?
Otherwise, if you're not careful about what thread you're in, you might be in a paho thread, and you might be blocking that thread while you're waiting, and that's not a good thing.

Reply:
I also assume that Timer.run will throw if you call cancel while it's running.  

---
In reply to: [277403544](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277403544) [](ancestors = 277403544)

Reply:
changed to calling trig_poll inside the time up function and start the timer. 



---
In reply to: [277407009](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277407009) [](ancestors = 277407009,277403544)